diff --git a/modules/ai_translate/ai_translate.services.yml b/modules/ai_translate/ai_translate.services.yml
index 36c511c1ff34777da6cd74d819867a66f19f3370..d4f9dbe2da1af02c1521c55e77892f1ee4a321c2 100644
--- a/modules/ai_translate/ai_translate.services.yml
+++ b/modules/ai_translate/ai_translate.services.yml
@@ -9,6 +9,15 @@ services:
       - '@entity_type.manager'
       - '@entity_field.manager'
       - '@plugin.manager.text_extractor'
+  ai_translate.text_translator:
+    class: Drupal\ai_translate\TextTranslator
+    arguments:
+      - '@entity_type.manager'
+      - '@language_manager'
+      - '@config.factory'
+      - '@ai.provider'
+      - '@twig'
+      - '@module_handler'
   plugin.manager.text_extractor:
     class: Drupal\ai_translate\FieldTextExtractorPluginManager
     parent: default_plugin_manager
diff --git a/modules/ai_translate/config/install/ai_translate.settings.yml b/modules/ai_translate/config/install/ai_translate.settings.yml
index 1714d254b936ead93a7abcb56a8c9cadb8b471dc..621e836b4ecfcedb4bea214d6c75ab37e6ebae47 100644
--- a/modules/ai_translate/config/install/ai_translate.settings.yml
+++ b/modules/ai_translate/config/install/ai_translate.settings.yml
@@ -1,7 +1,6 @@
 prompt: >1
   You are a helpful translator that can translate text and understand context when translating.
-  You will be given a context text to translate from the source language {{ source_lang_name }}
-  to the target language {{ dest_lang_name }}.
+  You will be given a context text to {% if source_lang_name %} translate from the source language {{ source_lang_name }} {% else %}guess the source language and translate {% endif %} to the target language {{ dest_lang_name }}.
   Only respond with the actual translation and nothing else.
   When translating the context text from the source language {{ source_lang_name }}
   to the target language {{ dest_lang_name }} take the following instructions into consideration:
diff --git a/modules/ai_translate/src/Controller/AiTranslateController.php b/modules/ai_translate/src/Controller/AiTranslateController.php
index 31d89ad3e9c27715168a46854edef80e55ea5d87..0a2e5aa35e381cc77184a612f8773c12ea763cfe 100644
--- a/modules/ai_translate/src/Controller/AiTranslateController.php
+++ b/modules/ai_translate/src/Controller/AiTranslateController.php
@@ -3,20 +3,16 @@
 namespace Drupal\ai_translate\Controller;
 
 use Drupal\Core\Batch\BatchBuilder;
-use Drupal\Core\Config\ImmutableConfig;
 use Drupal\Core\Controller\ControllerBase;
 use Drupal\Core\DependencyInjection\DependencySerializationTrait;
 use Drupal\Core\Entity\ContentEntityInterface;
 use Drupal\Core\Entity\EntityPublishedInterface;
 use Drupal\Core\Language\LanguageInterface;
-use Drupal\Core\Template\TwigEnvironment;
 use Drupal\Core\Url;
-use Drupal\ai\AiProviderPluginManager;
-use Drupal\ai\OperationType\Chat\ChatInput;
-use Drupal\ai\OperationType\Chat\ChatMessage;
 use Drupal\ai_translate\TextExtractorInterface;
+use Drupal\ai_translate\TextTranslatorInterface;
+use Drupal\ai_translate\TranslationException;
 use Drupal\filter\Entity\FilterFormat;
-use GuzzleHttp\Exception\GuzzleException;
 use Symfony\Component\DependencyInjection\ContainerInterface;
 use Symfony\Component\HttpFoundation\RedirectResponse;
 
@@ -27,27 +23,6 @@ class AiTranslateController extends ControllerBase {
 
   use DependencySerializationTrait;
 
-  /**
-   * AI module configuration.
-   *
-   * @var \Drupal\Core\Config\ImmutableConfig
-   */
-  protected ImmutableConfig $aiConfig;
-
-  /**
-   * AI provider plugin manager.
-   *
-   * @var \Drupal\ai\AiProviderPluginManager
-   */
-  protected AiProviderPluginManager $aiProviderManager;
-
-  /**
-   * Twig engine.
-   *
-   * @var \Drupal\Core\Template\TwigEnvironment
-   */
-  protected TwigEnvironment $twig;
-
   /**
    * Text extractor service.
    *
@@ -56,11 +31,11 @@ class AiTranslateController extends ControllerBase {
   protected TextExtractorInterface $textExtractor;
 
   /**
-   * The module handler for the hooks.
+   * Text translator service.
    *
-   * @var Drupal\Core\Extension\ModuleHandler
+   * @var \Drupal\ai_translate\TextTranslatorInterface
    */
-  protected $moduleHandler;
+  protected TextTranslatorInterface $aiTranslator;
 
   /**
    * {@inheritdoc}
@@ -69,11 +44,8 @@ class AiTranslateController extends ControllerBase {
     $instance = new static();
     $instance->entityTypeManager = $container->get('entity_type.manager');
     $instance->languageManager = $container->get('language_manager');
-    $instance->aiConfig = $container->get('config.factory')->get('ai.settings');
-    $instance->aiProviderManager = $container->get('ai.provider');
-    $instance->twig = $container->get('twig');
     $instance->textExtractor = $container->get('ai_translate.text_extractor');
-    $instance->moduleHandler = $container->get('module_handler');
+    $instance->aiTranslator = $container->get('ai_translate.text_translator');
     return $instance;
   }
 
@@ -139,96 +111,6 @@ class AiTranslateController extends ControllerBase {
     return batch_process($redirectUrl);
   }
 
-  /**
-   * Get the text translated by AI API call.
-   *
-   * @param string $input_text
-   *   Input prompt for the LLm.
-   * @param \Drupal\Core\Language\LanguageInterface $langFrom
-   *   Source language.
-   * @param \Drupal\Core\Language\LanguageInterface $langTo
-   *   Destination language.
-   *
-   * @return string
-   *   Translated content.
-   */
-  public function translateContent(
-    string $input_text,
-    LanguageInterface $langFrom,
-    LanguageInterface $langTo,
-  ) {
-    $preferred_model = $this->config('ai_translate.settings')->get($langTo->id() . '_model');
-    $provider_config = $this->getSetProvider($preferred_model, 'chat');
-    $provider = $provider_config['provider_id'];
-    $prompt = $this->config('ai_translate.settings')->get($langTo->id() . '_prompt');
-    if (empty($prompt)) {
-      $prompt = $this->config('ai_translate.settings')->get('prompt');
-    }
-    $promptText = $this->twig->renderInline($prompt, [
-      'source_lang' => $langFrom->getId(),
-      'source_lang_name' => $langFrom->getName(),
-      'dest_lang' => $langTo->getId(),
-      'dest_lang_name' => $langTo->getName(),
-      'input_text' => $input_text,
-    ]);
-    try {
-      $messages = new ChatInput([
-        new chatMessage('system', 'You are helpful translator. '),
-        new chatMessage('user', $promptText),
-      ]);
-
-      // Allow other modules to take over.
-      $this->moduleHandler->alter('ai_translate_translation', $messages, $provider, $provider_config['model_id']);
-
-      /** @var /Drupal\ai\OperationType\Chat\ChatOutput $message */
-      $message = $provider->chat($messages, $provider_config['model_id'], ['ai_translate'])->getNormalized();
-    }
-    catch (GuzzleException $exception) {
-      // Error handling for the API call.
-      return $exception->getMessage();
-    }
-    $cleaned = trim(trim($message->getText(), '```'), ' ');
-    return trim($cleaned, '"');
-  }
-
-  /**
-   * Get the preferred provider if configured, else take the default one.
-   *
-   * @param string $preferred_model
-   *   The preferred model as a string.
-   * @param string $operationType
-   *   The operation type (like chat).
-   *
-   * @return array|null
-   *   An array with the model and provider.
-   *
-   * @throws \Drupal\Component\Plugin\Exception\PluginException
-   *   An exception.
-   */
-  public function getSetProvider($preferred_model, $operationType) {
-    // Check if there is a preferred model.
-    $provider = NULL;
-    $model = NULL;
-    if ($preferred_model) {
-      $provider = $this->aiProviderManager->loadProviderFromSimpleOption($preferred_model);
-      $model = $this->aiProviderManager->getModelNameFromSimpleOption($preferred_model);
-    }
-    else {
-      // Get the default provider.
-      $default_provider = $this->aiProviderManager->getDefaultProviderForOperationType($operationType);
-      if (empty($default_provider['provider_id'])) {
-        // If we got nothing return NULL.
-        return NULL;
-      }
-      $provider = $this->aiProviderManager->createInstance($default_provider['provider_id']);
-      $model = $default_provider['model_id'];
-    }
-    return [
-      'provider_id' => $provider,
-      'model_id' => $model,
-    ];
-  }
-
   /**
    * Finished operation.
    */
@@ -258,9 +140,14 @@ class AiTranslateController extends ControllerBase {
     array &$context,
   ) {
     // Translate the content.
-    $translated_text = $this->translateContent(
-      $singleText['value'], $langFrom, $langTo
-    );
+    try {
+      $translated_text = $this->aiTranslator->translateContent(
+        $singleText['value'], $langTo, $langFrom);
+    }
+    catch (TranslationException) {
+      $context['results']['failures'][] = $singleText['value'];
+      return;
+    }
 
     // Checks if the field allows HTML and decodes the HTML entities.
     if (isset($singleText['format'])) {
diff --git a/modules/ai_translate/src/Drush/AiTranslateCommands.php b/modules/ai_translate/src/Drush/AiTranslateCommands.php
index acd6cdb0367576bc1592c01e91dfea8c5057ad65..20a72806bf5e84ec01d6efd920e3f7403f631e6b 100644
--- a/modules/ai_translate/src/Drush/AiTranslateCommands.php
+++ b/modules/ai_translate/src/Drush/AiTranslateCommands.php
@@ -7,8 +7,9 @@ use Drupal\Core\Language\LanguageManagerInterface;
 use Drupal\Core\Logger\LoggerChannelTrait;
 use Drupal\Core\Messenger\MessengerTrait;
 use Drupal\Core\StringTranslation\StringTranslationTrait;
-use Drupal\ai_translate\Controller\AiTranslateController;
 use Drupal\ai_translate\TextExtractorInterface;
+use Drupal\ai_translate\TextTranslatorInterface;
+use Drupal\ai_translate\TranslationException;
 use Drush\Attributes\Argument;
 use Drush\Attributes\Command;
 use Drush\Commands\DrushCommands;
@@ -45,11 +46,11 @@ class AiTranslateCommands extends DrushCommands {
   protected TextExtractorInterface $textExtractor;
 
   /**
-   * Translate controller.
+   * Text translation service.
    *
-   * @var \Drupal\ai_translate\Controller\AiTranslateController
+   * @var \Drupal\ai_translate\TextTranslatorInterface
    */
-  protected AiTranslateController $translateController;
+  protected TextTranslatorInterface $textTranslator;
 
   /**
    * {@inheritdoc}
@@ -59,8 +60,7 @@ class AiTranslateCommands extends DrushCommands {
     $instance->languageManager = $container->get('language_manager');
     $instance->entityTypeManager = $container->get('entity_type.manager');
     $instance->textExtractor = $container->get('ai_translate.text_extractor');
-    $instance->translateController = \Drupal::classResolver(AiTranslateController::class
-    );
+    $instance->textTranslator = $container->get('ai_translate.text_translator');
     return $instance;
   }
 
@@ -68,13 +68,13 @@ class AiTranslateCommands extends DrushCommands {
    * Create AI-powered translation of an entity.
    */
   #[Command(
-    name: 'ai:translate'
+    name: 'ai:translate-entity'
   )]
   #[Argument(name: 'entityType', description: 'Entity type (i.e. node)')]
   #[Argument(name: 'entityId', description: 'Entity ID (i.e. 16)')]
   #[Argument(name: 'langFrom', description: 'Source language code (i.e. fr)')]
   #[Argument(name: 'langTo', description: 'Target language code (i.e. en)')]
-  public function translate(
+  public function translateEntity(
     string $entityType,
     string $entityId,
     string $langFrom,
@@ -84,8 +84,6 @@ class AiTranslateCommands extends DrushCommands {
     if (empty($langNames)) {
       $langNames = $this->languageManager->getNativeLanguages();
     }
-    $langFromName = $langNames[$langFrom]->getName();
-    $langToName = $langNames[$langTo]->getName();
     $entity = $this->entityTypeManager->getStorage($entityType)
       ->load($entityId);
     if ($entity->language()->getId() !== $langFrom
@@ -99,8 +97,15 @@ class AiTranslateCommands extends DrushCommands {
     }
     $textMetadata = $this->textExtractor->extractTextMetadata($entity);
     foreach ($textMetadata as &$singleText) {
-      $singleText['translated'] = $this->translateController
-        ->translateContent($singleText['value'], $langFromName, $langToName);
+      try {
+        $singleText['translated'] = $this->textTranslator->translateContent(
+          $singleText['value'], $langNames[$langTo], $langNames[$langFrom] ?? NULL);
+      }
+      catch (TranslationException) {
+        // Error already logged by text_translate service.
+        $this->messenger()->addError('Error translating content.');
+        return;
+      }
     }
     $translation = $entity->addTranslation($langTo);
     $this->textExtractor->insertTextMetadata($translation,
@@ -116,4 +121,33 @@ class AiTranslateCommands extends DrushCommands {
 
   }
 
+  /**
+   * Create AI-powered translation of a text.
+   */
+  #[Command(
+    name: 'ai:translate-text'
+  )]
+  #[Argument(name: 'text', description: 'Text to translate')]
+  #[Argument(name: 'langTo', description: 'Target language code (i.e. en)')]
+  #[Argument(name: 'langFrom', description: 'Source language code (i.e. fr)')]
+  public function translate(
+    string $text,
+    string $langFrom,
+    string $langTo,
+  ) {
+    static $langNames;
+    if (empty($langNames)) {
+      $langNames = $this->languageManager->getNativeLanguages();
+    }
+    try {
+      return $this->textTranslator->translateContent($text,
+        $langNames[$langTo], $langNames[$langFrom] ?? NULL);
+    }
+    catch (TranslationException) {
+      // Error already logged by text_translate service.
+      $this->messenger()->addError('Error translating content.');
+      return;
+    }
+  }
+
 }
diff --git a/modules/ai_translate/src/Form/AiTranslateForm.php b/modules/ai_translate/src/Form/AiTranslateForm.php
index 2926722ccbf6a584f13e90b3cd51ae0c5fdf62aa..3a2c447a88d6b401c1b8e6d5bdf531cb7f3d5dba 100644
--- a/modules/ai_translate/src/Form/AiTranslateForm.php
+++ b/modules/ai_translate/src/Form/AiTranslateForm.php
@@ -133,6 +133,10 @@ class AiTranslateForm extends FormBase {
       if ($lang_from !== $langcode && !$entity->hasTranslation($langcode)) {
         $model = $config->get($langcode . '_model') ?? '';
         $parts = explode('__', $model);
+        if (empty($parts[0])) {
+          $default_model = $this->providerManager->getSimpleDefaultProviderOptions('chat');
+          $parts1 = explode('__', $default_model);
+          $ai_model = $parts1[1];
         if ($model == "" || empty($parts[0])) {
           $default_model = $this->providerManager->getSimpleDefaultProviderOptions('chat');
           if ($default_model == "") {
diff --git a/modules/ai_translate/src/Form/AiTranslateSettingsForm.php b/modules/ai_translate/src/Form/AiTranslateSettingsForm.php
index 44cf87d8f8cc2770b513921ec1731cf5e4b256f4..2256236386b32cdcc5b61aab3018e1f1c0560333 100644
--- a/modules/ai_translate/src/Form/AiTranslateSettingsForm.php
+++ b/modules/ai_translate/src/Form/AiTranslateSettingsForm.php
@@ -2,6 +2,7 @@
 
 namespace Drupal\ai_translate\Form;
 
+use Drupal\Core\Config\FileStorage;
 use Drupal\Core\Entity\ContentEntityTypeInterface;
 use Drupal\Core\Entity\EntityTypeManagerInterface;
 use Drupal\Core\Extension\ModuleHandlerInterface;
@@ -107,6 +108,12 @@ class AiTranslateSettingsForm extends ConfigFormBase {
     $example_prompt = $config->get('prompt');
 
     $languages = $this->languageManager->getLanguages();
+    $form['prompt'] = [
+      '#title' => $this->t('Default translation prompt'),
+      '#type' => 'textarea',
+      '#required' => TRUE,
+      '#default_value' => $example_prompt ?? '',
+    ];
     foreach ($languages as $langcode => $language) {
       $form[$langcode] = [
         '#type' => 'fieldset',
@@ -124,16 +131,22 @@ class AiTranslateSettingsForm extends ConfigFormBase {
       ];
       $form[$langcode]['prompt'] = [
         '#title' => $this->t('Translation prompt for translating to @lang', ['@lang' => $language->getName()]),
+        '#description' => $this->t('Leave empty to use the default translation prompt.'),
         '#type' => 'textarea',
-        '#required' => TRUE,
+        '#required' => FALSE,
         '#default_value' => $config->get($langcode . '_prompt') ?? $example_prompt,
       ];
     }
 
-    $example_prompt = '<h3>Example prompt</h3><pre> ' . $example_prompt . ' </pre>';
-    $form['example_prompt'] = [
-      '#type' => 'markup',
-      '#markup' => $example_prompt,
+    $moduleInfo = $this->moduleHandler->getModule('ai_translate');
+    $config_path = $moduleInfo->getPath() . '/config/install';
+    $source = new FileStorage($config_path);
+    $source->read('ai_translate.settings');
+    $form['default_prompt'] = [
+      '#type' => 'inline_template',
+      '#template' => '<h3>Prompt suggested by module maintainers:</h3>
+<pre>{{ settings.prompt }}</pre>',
+      '#context' => ['settings' => $source->read('ai_translate.settings')],
     ];
 
     $helpText = $this->moduleHandler->moduleExists('help')
@@ -195,15 +208,30 @@ class AiTranslateSettingsForm extends ConfigFormBase {
    * {@inheritdoc}
    */
   public function validateForm(array &$form, FormStateInterface $form_state) {
-    $languages = $this->languageManager->getLanguages();
-    foreach ($languages as $langcode => $language) {
+    try {
+      if (strlen($this->twig->renderInline($form_state->getValue('prompt'), [
+        'source_lang_name' => 'Test 1',
+        'dest_lang_name' => 'Test 2',
+        'input_text' => 'Text to translate',
+      ])) < self::MINIMAL_PROMPT_LENGTH) {
+        $form_state->setErrorByName('prompt',
+          $this->t('Prompt cannot be shorter than @num characters',
+            ['@num' => self::MINIMAL_PROMPT_LENGTH]));
+      }
+    }
+    catch (\Exception $e) {
+      $form_state->setErrorByName('prompt', $e->getMessage());
+    }
+    foreach ($this->languageManager->getLanguages() as $langcode => $language) {
       try {
-        if (strlen($this->twig->renderInline($form_state->getValue($langcode)['prompt'], [
+        // Language-specific prompts are optional.
+        $langPrompt = $form_state->getValue([$langcode, 'prompt']);
+        if ($langPrompt && strlen($this->twig->renderInline($langPrompt, [
           'source_lang_name' => 'Test 1',
           'dest_lang_name' => 'Test 2',
           'input_text' => 'Text to translate',
         ])) < self::MINIMAL_PROMPT_LENGTH) {
-          $form_state->setErrorByName('prompt',
+          $form_state->setError($form[$langcode]['prompt'],
             $this->t('Prompt cannot be shorter than @num characters',
               ['@num' => self::MINIMAL_PROMPT_LENGTH]));
         }
@@ -224,8 +252,8 @@ class AiTranslateSettingsForm extends ConfigFormBase {
     $config->set('reference_defaults', array_keys(array_filter($form_state->getValue('reference_defaults'))));
     $languages = $this->languageManager->getLanguages();
     foreach ($languages as $langcode => $language) {
-      $config->set($langcode . '_model', $form_state->getValue($langcode)['model']);
-      $config->set($langcode . '_prompt', $form_state->getValue($langcode)['prompt']);
+      $config->set($langcode . '_model', $form_state->getValue([$langcode, 'model']));
+      $config->set($langcode . '_prompt', $form_state->getValue([$langcode, 'prompt']));
     }
     $config->save();
     parent::submitForm($form, $form_state);
diff --git a/modules/ai_translate/src/Plugin/AiProvider/ChatTranslationProvider.php b/modules/ai_translate/src/Plugin/AiProvider/ChatTranslationProvider.php
new file mode 100644
index 0000000000000000000000000000000000000000..d93b4afad556c133ecbf8433875847b9a8afaf02
--- /dev/null
+++ b/modules/ai_translate/src/Plugin/AiProvider/ChatTranslationProvider.php
@@ -0,0 +1,229 @@
+<?php
+
+namespace Drupal\ai_translate\Plugin\AiProvider;
+
+use Drupal\Core\Config\ImmutableConfig;
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\Core\StringTranslation\TranslatableMarkup;
+use Drupal\Core\Template\TwigEnvironment;
+use Drupal\ai\AiProviderPluginManager;
+use Drupal\ai\Attribute\AiProvider;
+use Drupal\ai\Base\AiProviderClientBase;
+use Drupal\ai\Enum\AiProviderCapability;
+use Drupal\ai\OperationType\Chat\ChatInput;
+use Drupal\ai\OperationType\Chat\ChatMessage;
+use Drupal\ai\OperationType\TranslateText\TranslateTextInput;
+use Drupal\ai\OperationType\TranslateText\TranslateTextInterface;
+use Drupal\ai\OperationType\TranslateText\TranslateTextOutput;
+use Drupal\ai\Plugin\ProviderProxy;
+use Drupal\language\Entity\ConfigurableLanguage;
+use GuzzleHttp\Exception\GuzzleException;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Plugin implementation of the 'chat_translation' provider.
+ *
+ * The purpose is to implement 'translate_text' operation
+ * using more generic 'chat' operation supported by many LLM providers.
+ */
+#[AiProvider(
+  id: 'chat_translation',
+  label: new TranslatableMarkup('Chat proxy to LLM'),
+)]
+class ChatTranslationProvider extends AiProviderClientBase implements
+  ContainerFactoryPluginInterface,
+  TranslateTextInterface {
+
+  use StringTranslationTrait;
+
+  /**
+   * AI provider plugin manager.
+   *
+   * @var \Drupal\ai\AiProviderPluginManager
+   */
+  protected AiProviderPluginManager $manager;
+
+  /**
+   * Configuration of default chat provider.
+   *
+   * @var array
+   */
+  protected array $chatConfiguration;
+
+  /**
+   * Lazy-loaded provider that actually performs text translation.
+   *
+   * @var \Drupal\ai\Plugin\ProviderProxy
+   */
+  protected ProviderProxy $realTranslator;
+
+  /**
+   * The Twig engine.
+   *
+   * @var \Drupal\Core\Template\TwigEnvironment
+   */
+  protected TwigEnvironment $twig;
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    $instance = parent::create($container, $configuration, $plugin_id, $plugin_definition);
+    $instance->manager = $container->get('ai.provider');
+    $instance->twig = $container->get('twig');
+    return $instance;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function isUsable(?string $operation_type = NULL, array $capabilities = []): bool {
+    if (empty($this->configFactory->get('ai_translate.settings')->get('prompt'))) {
+      return FALSE;
+    }
+    if (!isset($this->chatConfiguration)) {
+      $defaultProviders = $this->config->get('default_providers');
+      if (empty($defaultProviders)) {
+        $this->chatConfiguration = [];
+      }
+      $this->chatConfiguration = $defaultProviders['chat'] ?? [];
+    }
+    return !empty($this->chatConfiguration)
+      && in_array($operation_type, $this->getSupportedOperationTypes());
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getSupportedOperationTypes(): array {
+    return [
+      'translate_text',
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getSupportedCapabilities(): array {
+    return [
+      AiProviderCapability::StreamChatOutput,
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getConfig(): ImmutableConfig {
+    return $this->configFactory->get('ai.settings');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getConfiguredModels(?string $operation_type = NULL, array $capabilities = []): array {
+    return ['default' => $this->t('default')];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getApiDefinition(): array {
+    return [];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getModelSettings(string $model_id, array $generalConfig = []): array {
+    // This provider does not have model settings.
+    return [];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setAuthentication(mixed $authentication): void {}
+
+  /**
+   * {@inheritdoc}
+   */
+  public function translateText(TranslateTextInput $input, string $model_id, array $options = []): TranslateTextOutput {
+    $text = $input->getText();
+
+    // We can guess source, but not target language.
+    $targetLanguage = ConfigurableLanguage::load($input->getTargetLanguage());
+    if (!$targetLanguage) {
+      // @todo TranslateText-specific exception, documented in
+      // TranslateTextInterface::translateText() docblock.
+      $this->loggerFactory->get('ai_translate')->warning(
+        $this->t('Unable to guess target language, code @langcode',
+          ['@langcode' => $input->getTargetLanguage()]));
+      return new TranslateTextOutput('', '', '');
+    }
+
+    $aiConfig = $this->configFactory->get('ai_translate.settings');
+    $prompt = $aiConfig->get($targetLanguage->getId() . '_prompt');
+    if (empty($prompt)) {
+      $prompt = $aiConfig->get('prompt');
+    }
+    $context = [
+      'dest_lang' => $targetLanguage->getId(),
+      'dest_lang_name' => $targetLanguage->getName(),
+      'input_text' => $text,
+    ];
+    try {
+      $sourceLanguage = ConfigurableLanguage::load($input->getSourceLanguage());
+      if ($sourceLanguage) {
+        $context['source_lang'] = $sourceLanguage->getId();
+        $context['source_lang_name'] = $sourceLanguage->getName();
+      }
+    }
+    // Ignore failure to load source language.
+    catch (\AssertionError) {
+    }
+    $promptText = $this->twig->renderInline($prompt, $context);
+    try {
+      $this->setChatSystemRole('You are a helpful translator.');
+      $messages = new ChatInput([
+        new chatMessage('user', $promptText),
+      ]);
+
+      // Allow other modules to take over.
+      $this->moduleHandler->alter('ai_translate_translation', $messages, $provider, $provider_config['model_id']);
+
+      $this->loadTranslator();
+      /** @var /Drupal\ai\OperationType\Chat\ChatOutput $message */
+      $message = $this->realTranslator->chat($messages, $this->chatConfiguration['model_id']);
+    }
+    catch (GuzzleException $exception) {
+      // Error handling for the API call.
+      $this->loggerFactory->get('ai_translate')
+        ->warning($exception->getMessage());
+      return new TranslateTextOutput('', '', '');
+    }
+
+    return new TranslateTextOutput($message->getNormalized()->getText(),
+      $message->getRawOutput(), []);
+  }
+
+  /**
+   * Load real translator and its configuration.
+   *
+   * @return \Drupal\ai\Plugin\ProviderProxy|null
+   *   Real provider or NULL on failure.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\PluginException
+   */
+  protected function loadTranslator() :? ProviderProxy {
+    if (isset($this->realTranslator)) {
+      return $this->realTranslator;
+    }
+    $this->chatConfiguration = $this->manager
+      ->getDefaultProviderForOperationType('chat');
+    $this->realTranslator = $this->manager->createInstance(
+      $this->chatConfiguration['provider_id'], $this->chatConfiguration);
+    return $this->realTranslator;
+  }
+
+}
diff --git a/modules/ai_translate/src/TextTranslator.php b/modules/ai_translate/src/TextTranslator.php
new file mode 100644
index 0000000000000000000000000000000000000000..1d344d170da1b42071ef1a51e6b141cd96d76316
--- /dev/null
+++ b/modules/ai_translate/src/TextTranslator.php
@@ -0,0 +1,68 @@
+<?php
+
+namespace Drupal\ai_translate;
+
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Config\ImmutableConfig;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Extension\ModuleHandlerInterface;
+use Drupal\Core\Language\LanguageInterface;
+use Drupal\Core\Language\LanguageManagerInterface;
+use Drupal\Core\Logger\LoggerChannelTrait;
+use Drupal\Core\Template\TwigEnvironment;
+use Drupal\ai\AiProviderPluginManager;
+use Drupal\ai\OperationType\TranslateText\TranslateTextInput;
+
+/**
+ * Defines text translator service.
+ */
+class TextTranslator implements TextTranslatorInterface {
+
+  use LoggerChannelTrait;
+
+  /**
+   * AI module configuration.
+   *
+   * @var \Drupal\Core\Config\ImmutableConfig
+   */
+  protected ImmutableConfig $aiConfig;
+
+  public function __construct(
+    protected EntityTypeManagerInterface $entityTypeManager,
+    protected LanguageManagerInterface $languageManager,
+    protected ConfigFactoryInterface $configFactory,
+    protected AiProviderPluginManager $aiProviderManager,
+    protected TwigEnvironment $twig,
+    protected ModuleHandlerInterface $moduleHandler,
+  ) {
+    $this->aiConfig = $this->configFactory->get('ai.settings');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function translateContent(
+    string $input_text,
+    LanguageInterface $langTo,
+    ?LanguageInterface $langFrom = NULL,
+    array $context = [],
+  ) : string {
+    try {
+      /** @var \Drupal\ai\OperationType\TranslateText\TranslateTextInterface $provider */
+      $providerConfig = $this->aiProviderManager->getDefaultProviderForOperationType('translate_text');
+      $provider = $this->aiProviderManager->createInstance($providerConfig['provider_id'], $providerConfig);
+      $translation = $provider->translateText(
+        new TranslateTextInput($input_text, $langFrom->getId(), $langTo->getId()),
+        'default'
+      );
+
+      $cleaned = trim(trim($translation->getNormalized(), '```'), ' ');
+      return trim($cleaned, '"');
+    }
+    catch (\Throwable $e) {
+      $this->getLogger('ai_translate')->warning($e->getMessage());
+      throw new TranslationException($e->getMessage());
+    }
+  }
+
+}
diff --git a/modules/ai_translate/src/TextTranslatorInterface.php b/modules/ai_translate/src/TextTranslatorInterface.php
new file mode 100644
index 0000000000000000000000000000000000000000..94b8322cb0ee89d5ca8469ace2c114160a4fed5b
--- /dev/null
+++ b/modules/ai_translate/src/TextTranslatorInterface.php
@@ -0,0 +1,41 @@
+<?php
+
+namespace Drupal\ai_translate;
+
+use Drupal\Core\Language\LanguageInterface;
+
+/**
+ * Defines text translator interface.
+ */
+interface TextTranslatorInterface {
+
+  /**
+   * Get the text translated by AI API call.
+   *
+   * @param string $input_text
+   *   Input prompt for the LLm.
+   * @param \Drupal\Core\Language\LanguageInterface $langTo
+   *   Destination language.
+   * @param \Drupal\Core\Language\LanguageInterface|null $langFrom
+   *   Source language.
+   * @param array $context
+   *   Translation context. Possible keys include, but not limited to:
+   *   - preferred_model
+   *   - preferred_version
+   *   - preferred_provider
+   *
+   *   Lang from is an optional argument, thus the order of to/lang arguments.
+   *
+   * @return string
+   *   Translated content.
+   *
+   * @throws \Drupal\ai_translate\TranslationException
+   */
+  public function translateContent(
+    string $input_text,
+    LanguageInterface $langTo,
+    ?LanguageInterface $langFrom = NULL,
+    array $context = [],
+  ) : string;
+
+}
diff --git a/modules/ai_translate/src/TranslationException.php b/modules/ai_translate/src/TranslationException.php
new file mode 100644
index 0000000000000000000000000000000000000000..250624fab03205e569ce9cc909a9135ad86795d8
--- /dev/null
+++ b/modules/ai_translate/src/TranslationException.php
@@ -0,0 +1,8 @@
+<?php
+
+namespace Drupal\ai_translate;
+
+/**
+ * Defines translation exception.
+ */
+class TranslationException extends \Exception {}
